README
NAME
    Date::Lectionary - Readings for the Christian Lectionary

VERSION
    Version 1.20180418

SYNOPSIS
        use Time::Piece;
        use Date::Lectionary;

        my $epiphany = Date::Lectionary->new('date'=>Time::Piece->strptime("2017-01-06", "%Y-%m-%d"));
        say $epiphany->day->name; #String representation of the name of the day in the liturgical calendar; e.g. 'The Epiphany'
        say $epiphany->year->name; #String representation of the name of the liturgical year; e.g. 'A'
        say ${$epiphany->readings}[0] #String representation of the first reading for the day.

DESCRIPTION
    Date::Lectionary takes a Time::Piece date and returns the liturgical day
    and associated readings for the day.

  ATTRIBUTES
   date
    The Time::Piece object date given at object construction.

   lectionary
    An optional attribute given at object creation time. Valid values are
    'acna' for the Anglican Church of North America lectionary and 'rcl' for
    the Revised Common Lectionary with complementary readings in ordinary
    time. This attribute defaults to 'acna' if no value is given.

   day
    A Date::Lectionary::Day object containing attributes related to the
    liturgical day.

    "type": Stores the type of liturgical day. 'fixedFeast' is returned for
    non-moveable feast days such as Christmas Day. 'moveableFeast' is
    returned for moveable feast days. Moveable feasts move to a Monday when
    they occure on a Sunday. 'Sunday' is returned for non-fixed feast
    Sundays of the liturgical year. 'noLect' is returned for days with no
    feast day or Sunday readings.

    "name": The name of the day in the lectionary. For noLect days a String
    representation of the day is returned as the name.

    "alt": The alternative name --- if one is given --- of the day in the
    lectionary. If there is no alternative name for the day, then the empty
    string will be returned.

    "multiLect": Returns 'yes' if the day has multiple services with
    readings associated with it. (E.g. Christmas Day, Easter, etc.) Returns
    'no' if the day is a normal lectioanry day with only one service and one
    set of readings.

   year
    A Date::Lectionary::Year object containing attributes related to the
    liturgical year the date given at object construction resides in.

    "name": Returns 'A', 'B', or 'C' depending on the liturgical year the
    date given at object construction resides in.

   readings
    Return an ArrayRef of the String representation of the day's readings if
    there are any. Readings in the ArrayRef are ordered in the array
    according to the order the readings are given in the lectionary. If
    mutliple readings exist for the day, an ArrayRef of HashRefs will be
    given.

      my $singleReading = Date::Lectionary->new(
          'date'       => Time::Piece->strptime( "2016-11-13", "%Y-%m-%d" ),
          'lectionary' => 'acna'
      );

      say ${ $testReading->readings }[1]; #Will print 'Ps 98', the second reading for the Sunday closest to November 16 in the default ACNA lectionary for year C.
      say $testReading->day->multiLect; #Will print 'no' because this day does not have multiple services in the lectionary.

      my $multiReading = Date::Lectionary->new(
          'date'       => Time::Piece->strptime( "2016-12-25", "%Y-%m-%d" ),
          'lectionary' => 'rcl'
      );

      say $multiReading->day->multiLect; #Will print 'yes' because this day does have multiple services in the lectionary.
      say ${ $multiReading->readings }[0]{name}; #Will print 'Christmas, Proper I', the first services of Christmas Day in the RCL
      say ${ $multiReading->readings }[1]{readings}[0]; #Will print 'Isaiah 62:6-12', the first reading of the second service 'Christmas, Proper II' on Christmas Day in the RCL.

SUBROUTINES/METHODS
  BUILD
    Constructor for the Date::Lectionary object. Takes a Time::Piect object,
    "date", to create the object.

  _buildMultiReadings
    Private method that returns an ArrayRef of HashRefs for the multiple
    services and lectionary readings associated with the date.

  _buildReadings
    Private method that returns an ArrayRef of strings for the lectionary
    readings associated with the date.

  _determineAdvent
    Private method that takes a Time::Piece date object to returns a
    Date::Advent object containing the dates for Advent of the current
    liturgical year.

AUTHOR
    Michael Wayne Arnold, "<michael at rnold.info>"

BUGS
    Please report any bugs or feature requests to "bug-date-lectionary at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Date-Lectionary>. I will
    be notified, and then you'll automatically be notified of progress on
    your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Date::Lectionary

    You can also look for information at:

    *   RT: CPAN's request tracker (report bugs here)

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Date-Lectionary>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/Date-Lectionary>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/Date-Lectionary>

    *   Search CPAN

        <http://search.cpan.org/dist/Date-Lectionary/>

ACKNOWLEDGEMENTS
    Many thanks to my beautiful wife, Jennifer, my amazing daughter,
    Rosemary, and my sweet son, Oliver. But, above all, SOLI DEO GLORIA!

LICENSE
    Copyright 2016-2018 MICHAEL WAYNE ARNOLD

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions are
    met:

    1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.

    2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
    IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
    TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
    PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
    HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
    SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
    TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
    PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
    LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
    NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



CHANGES
Revision history for Date::Lectionary

1.20180313 2018-03-13

 - Switched from Try::Tiny to Try::Catch & Try::Tiny::Tiny

1.20180314 2018-03-14

 - Corrected issue with Manifest file
 - Corrected issue with tarball

1.20180410 2018-04-10

 - Added author tests to ensure better distribution quality
 - Tidied up the code a little bit

1.20180410 2018-04-16

 - Corrected the look-up value for the Fourth Sunday of Easter in the ACNA Year B lectionary to resolve an issue for 2018-04-22 was not returning any readings

Makefile.PL
use v5.22;

use strict;
use warnings;
use ExtUtils::MakeMaker;
use File::ShareDir::Install;

install_share dist => 'share';

WriteMakefile(
    NAME               => 'Date::Lectionary',
    AUTHOR             => q{Michael Wayne Arnold <michael@rnold.info>},
    VERSION_FROM       => 'lib/Date/Lectionary.pm',
    ABSTRACT           => 'Date::Lectionary - Readings for the Christian Lectionary',
    LICENSE            => 'freebsd',
    PL_FILES           => {},
    MIN_PERL_VERSION   => 5.022000,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker'     => 0,
        'File::ShareDir::Install' => '0.11'
    },
    BUILD_REQUIRES => {
        'Test::More'            => '1.302040',
        'Test::DistManifest'    => '1.014',
        'Test::Exception'       => '0.43',
        'Pod::Markdown'         => '3.005',
        'Test::Pod'             => '1.51',
        'Test::Pod::Coverage'   => '1.10',
        'Test::MinimumVersion'  => '0.101082',
        'Test::Kwalitee::Extra' => '0.4.0',
        'Test::Kwalitee'        => '1.27',
        'Test::CPAN::Changes'   => '0.400002',
        'Test::Version'         => '2.07',
    },
    PREREQ_PM => {
        'Carp'                         => '1.38',
        'Try::Tiny::Tiny'              => '0.001',
        'Try::Catch'                   => '1.1.0',
        'Time::Piece'                  => '1.3204',
        'Time::Seconds'                => '1.3204',
        'Moose'                        => '2.2010',
        'Moose::Util::TypeConstraints' => '2.2010',
        'MooseX::Aliases'              => '0.11',
        'Date::Advent'                 => '1.20171223',
        'Date::Easter'                 => '1.22',
        'Date::Lectionary::Time'       => '1.20170311',
        'namespace::autoclean'         => '0.28',
        'File::Share'                  => '0.25',
        'XML::LibXML'                  => '2.0128',
        'MooseX::StrictConstructor'    => '0.21',
    },
    META_MERGE => {
        "meta-spec" => { version => 2 },
        resources   => {
            repository => {
                type => 'git',
                url  => 'git://github.com/marmanold/Date-Lectionary.git',
                web  => 'https://github.com/marmanold/Date-Lectionary',
            },
        },
        provides => {
            'Date::Lectionary' => {
                file    => 'lib/Date/Lectionary.pm', 
                version => '1.20180418'
            },
            'Date::Lectionary::Day' => {
                file    => 'lib/Date/Lectionary/Day.pm', 
                version => '1.20180418'
            },
            'Date::Lectionary::Year' => {
                file    => 'lib/Date/Lectionary/Year.pm', 
                version => '1.20180418'
            },
        },
        keywords  => [qw/ lectionary rcl acna readings christian liturgical /],
        x_twitter => 'https://www.twitter.com/marmanold/',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES    => 'Date-Lectionary-*' },
);

package MY;
use File::ShareDir::Install qw(postamble);


lib/Date/Lectionary/Year.pm
package Date::Lectionary::Year;

use v5.22;
use strict;
use warnings;

use Try::Tiny::Tiny;

use Moose;
use MooseX::StrictConstructor;
use Carp;
use Try::Catch;
use namespace::autoclean;
use Moose::Util::TypeConstraints;

=head1 NAME

Date::Lectionary::Year - Cycle Year for the Lectionary

=head1 VERSION

Version 1.20180418

=cut

use version; our $VERSION = version->declare("v1.20180418");

=head1 SYNOPSIS

A helper object for Date::Lectionary to package which keeps information about the liturgical cycle year for the RCL, ACNA, and possibly other three-year lectionary systems.  Valid values for the liturgical cycle are A, B, or C.

=cut

enum 'litCycleYear', [qw(A B C)];
no Moose::Util::TypeConstraints;

=head1 SUBROUTINES/METHODS

=cut

has 'year' => (
    is       => 'ro',
    isa      => 'Int',
    required => 1,
);

has 'name' => (
    is       => 'ro',
    isa      => 'litCycleYear',
    writer   => '_setName',
    init_arg => undef,
);

=head2 BUILD

Constructor for the Date::Lectionary::Year object.  Takes a four-digit representation of the Common Era year and returns the correct liturgical cycle year for the RCL, ACNA, and possibly other three-year lectionary systems.

=cut

sub BUILD {
    my $self = shift;

    $self->_setName( _determineYear( $self->year ) );
}

=head2 _determineYear

Private method that takes a four-digit representation of the Common Era year and calculates the liturgical year -- A, B, or C -- for the year.

=cut

sub _determineYear {
    my $calYear = shift;

    try {
        if ( $calYear % 3 == 0 ) {
            return 'A';
        }
        elsif ( ( $calYear - 1 ) % 3 == 0 ) {
            return 'B';
        }
        elsif ( ( $calYear - 2 ) % 3 == 0 ) {
            return 'C';
        }
        else {
            confess "The liturgical year for the year [" . $calYear . "] could not be determined.";
        }
    }
    catch {
        confess "A liturgical year for the value [" . $calYear . "] could not be calculated.";
    };
}

=head1 AUTHOR

Michael Wayne Arnold, C<< <michael at rnold.info> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-date-lectionary at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Date-Lectionary-Year>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Date::Lectionary::Year


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Date-Lectionary-Year>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Date-Lectionary-Year>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Date-Lectionary-Year>

=item * Search CPAN

L<http://search.cpan.org/dist/Date-Lectionary-Year/>

=back

=head1 ACKNOWLEDGEMENTS

Many thanks to my beautiful wife, Jennifer, my amazing daughter, Rosemary, and my sweet son, Oliver.  But, above all, SOLI DEO GLORIA!

=head1 LICENSE

Copyright 2016-2018 MICHAEL WAYNE ARNOLD

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

__PACKAGE__->meta->make_immutable;

1;    # End of Date::Lectionary::Year


lib/Date/Lectionary/Reading.pm
package Date::Lectionary::Reading;

use v5.22;
use strict;
use warnings;

use Moose;
use Carp;
use Try::Tiny;
use namespace::autoclean;
use Moose::Util::TypeConstraints;

=head1 NAME

Date::Lectionary::Reading

=head1 VERSION

Version 1.20160820

=cut

our $VERSION = '1.20160820';

=head1 SYNOPSIS

A helper object for Date::Lectionary to package together all the parts of a scripture reading.

=cut

enum 'Testament',   [qw(OT NT)];
enum 'ReadingType', [qw(OT Psalm Epistle Gospel)];
no Moose::Util::TypeConstraints;

=head1 SUBROUTINES/METHODS

=cut

has 'book' => (
    is       => 'ro',
    isa      => 'Str',
    required => 1,
);

has 'begin' => (
    is       => 'ro',
    isa      => 'Str',
    required => 1,
);

has 'end' => (
    is       => 'ro',
    isa      => 'Str',
    required => 1,
);

has 'testament' => (
    is      => 'ro',
    isa     => 'Testament',
    builder => '_buildTestament',
);

has 'type' => (
    is      => 'ro',
    isa     => 'ReadingType',
    builder => '_buildType'
);

=head2 _buildTestament

Private method that determines which testament the given reading's book is in.

=cut

sub _buildTestament {
    my $self = shift;

    return 'OT';
}

=head2 _buildType

Private method that determines what type of reading the given reading is.

=cut

sub _buildType {
    my $self = shift;

    return 'OT';
}

=head1 AUTHOR

Michael Wayne Arnold, C<< <marmanold at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-date-lectionary at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Date-Lectionary-Reading>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Date::Lectionary


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Date-Lectionary-Reading>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Date-Lectionary-Reading>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Date-Lectionary-Reading>

=item * Search CPAN

L<http://search.cpan.org/dist/Date-Lectionary-Reading/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2016 Michael Wayne Arnold.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.


=cut

__PACKAGE__->meta->make_immutable;

1;    # End of Date::Lectionary::Reading


lib/Date/Lectionary.pm
package Date::Lectionary;

use v5.22;
use strict;
use warnings;

use Try::Tiny::Tiny;

use Moose;
use MooseX::StrictConstructor;
use Carp;
use Try::Catch;
use XML::LibXML;
use File::Share ':all';
use Time::Piece;
use Date::Advent;
use Date::Lectionary::Year;
use Date::Lectionary::Day;
use namespace::autoclean;
use Moose::Util::TypeConstraints;

=head1 NAME

Date::Lectionary - Readings for the Christian Lectionary

=head1 VERSION

Version 1.20180418

=cut

use version; our $VERSION = version->declare("v1.20180418");

=head1 SYNOPSIS

    use Time::Piece;
    use Date::Lectionary;

    my $epiphany = Date::Lectionary->new('date'=>Time::Piece->strptime("2017-01-06", "%Y-%m-%d"));
    say $epiphany->day->name; #String representation of the name of the day in the liturgical calendar; e.g. 'The Epiphany'
    say $epiphany->year->name; #String representation of the name of the liturgical year; e.g. 'A'
    say ${$epiphany->readings}[0] #String representation of the first reading for the day.

=head1 DESCRIPTION

Date::Lectionary takes a Time::Piece date and returns the liturgical day and associated readings for the day.

=head2 ATTRIBUTES

=head3 date

The Time::Piece object date given at object construction.

=head3 lectionary

An optional attribute given at object creation time.  Valid values are 'acna' for the Anglican Church of North America lectionary and 'rcl' for the Revised Common Lectionary with complementary readings in ordinary time.  This attribute defaults to 'acna' if no value is given.

=head3 day

A Date::Lectionary::Day object containing attributes related to the liturgical day.

C<type>: Stores the type of liturgical day. 'fixedFeast' is returned for non-moveable feast days such as Christmas Day. 'moveableFeast' is returned for moveable feast days.  Moveable feasts move to a Monday when they occure on a Sunday. 'Sunday' is returned for non-fixed feast Sundays of the liturgical year.  'noLect' is returned for days with no feast day or Sunday readings.

C<name>: The name of the day in the lectionary.  For noLect days a String representation of the day is returned as the name.

C<alt>: The alternative name --- if one is given --- of the day in the lectionary.  If there is no alternative name for the day, then the empty string will be returned.

C<multiLect>: Returns 'yes' if the day has multiple services with readings associated with it.  (E.g. Christmas Day, Easter, etc.)  Returns 'no' if the day is a normal lectioanry day with only one service and one set of readings.

=head3 year

A Date::Lectionary::Year object containing attributes related to the liturgical year the date given at object construction resides in.

C<name>: Returns 'A', 'B', or 'C' depending on the liturgical year the date given at object construction resides in.

=head3 readings

Return an ArrayRef of the String representation of the day's readings if there are any.  Readings in the ArrayRef are ordered in the array according to the order the readings are given in the lectionary.  If mutliple readings exist for the day, an ArrayRef of HashRefs will be given.

  my $singleReading = Date::Lectionary->new(
      'date'       => Time::Piece->strptime( "2016-11-13", "%Y-%m-%d" ),
      'lectionary' => 'acna'
  );

  say ${ $testReading->readings }[1]; #Will print 'Ps 98', the second reading for the Sunday closest to November 16 in the default ACNA lectionary for year C.
  say $testReading->day->multiLect; #Will print 'no' because this day does not have multiple services in the lectionary.

  my $multiReading = Date::Lectionary->new(
      'date'       => Time::Piece->strptime( "2016-12-25", "%Y-%m-%d" ),
      'lectionary' => 'rcl'
  );

  say $multiReading->day->multiLect; #Will print 'yes' because this day does have multiple services in the lectionary.
  say ${ $multiReading->readings }[0]{name}; #Will print 'Christmas, Proper I', the first services of Christmas Day in the RCL
  say ${ $multiReading->readings }[1]{readings}[0]; #Will print 'Isaiah 62:6-12', the first reading of the second service 'Christmas, Proper II' on Christmas Day in the RCL.

=cut

enum 'LectionaryType', [qw(acna rcl)];
no Moose::Util::TypeConstraints;

=head1 SUBROUTINES/METHODS

=cut

has 'date' => (
    is       => 'ro',
    isa      => 'Time::Piece',
    required => 1,
);

has 'day' => (
    is       => 'ro',
    isa      => 'Date::Lectionary::Day',
    writer   => '_setDay',
    init_arg => undef,
);

has 'year' => (
    is       => 'ro',
    isa      => 'Date::Lectionary::Year',
    writer   => '_setYear',
    init_arg => undef,
);

has 'lectionary' => (
    is      => 'ro',
    isa     => 'LectionaryType',
    default => 'acna',
);

has 'readings' => (
    is       => 'ro',
    isa      => 'ArrayRef',
    writer   => '_setReadings',
    init_arg => undef,
);

=head2 BUILD

Constructor for the Date::Lectionary object.  Takes a Time::Piect object, C<date>, to create the object.

=cut

sub BUILD {
    my $self = shift;

    my $advent = _determineAdvent( $self->date );

    $self->_setYear( Date::Lectionary::Year->new( 'year' => $advent->firstSunday->year ) );

    $self->_setDay(
        Date::Lectionary::Day->new(
            'date'       => $self->date,
            'lectionary' => $self->lectionary
        )
    );

    if ( $self->day->multiLect eq 'yes' ) {
        $self->_setReadings( _buildMultiReadings( $self->day->subLects, $self->lectionary, $self->year->name ) );
    }
    else {
        $self->_setReadings( _buildReadings( $self->day->name, $self->lectionary, $self->year->name ) );
    }
}

=head2 _buildMultiReadings

Private method that returns an ArrayRef of HashRefs for the multiple services and lectionary readings associated with the date.

=cut

sub _buildMultiReadings {
    my $multiNames = shift;
    my $lectionary = shift;
    my $year       = shift;

    my @multiReadings;
    foreach my $name (@$multiNames) {

        my %lectPart = (
            name     => $name,
            readings => _buildReadings( $name, $lectionary, $year )
        );

        push( @multiReadings, \%lectPart );
    }

    return \@multiReadings;
}

=head2 _buildReadings

Private method that returns an ArrayRef of strings for the lectionary readings associated with the date.

=cut

sub _buildReadings {
    my $displayName = shift;
    my $lectionary  = shift;
    my $year        = shift;

    my $parser = XML::LibXML->new();
    my $data_location;
    my $readings;

    try {
        $data_location = dist_file( 'Date-Lectionary', $lectionary . '_lect.xml' );
        $readings = $parser->parse_file($data_location);
    }
    catch {
        carp "The readings database for the $lectionary lectionary could not be found or parsed.";
    };

    my $compiled_xpath = XML::LibXML::XPathExpression->new("/lectionary/year[\@name=\"$year\" or \@name=\"holidays\"]/day[\@name=\"$displayName\"]/lesson");

    my @readings;
    try {
        foreach my $lesson ( $readings->findnodes($compiled_xpath) ) {
            push( @readings, $lesson->to_literal );
        }
    }
    catch {
        carp "Readings for $displayName in year $year could not be parsed from the database.";
    };

    return \@readings;
}

=head2 _determineAdvent

Private method that takes a Time::Piece date object to returns a Date::Advent object containing the dates for Advent of the current liturgical year.

=cut

sub _determineAdvent {
    my $date = shift;

    my $advent = undef;

    try {
        $advent = Date::Advent->new( date => $date );
        return $advent;
    }
    catch {
        confess "Could not calculate Advent for the given date [" . $date->ymd . "].";
    };
}

=head1 AUTHOR

Michael Wayne Arnold, C<< <michael at rnold.info> >>

=head1 BUGS

=for html <a href="https://travis-ci.org/marmanold/Date-Lectionary"><img src="https://travis-ci.org/marmanold/Date-Lectionary.svg?branch=master"></a>

=for html <a href='https://coveralls.io/github/marmanold/Date-Lectionary?branch=master'><img src='https://coveralls.io/repos/github/marmanold/Date-Lectionary/badge.svg?branch=master' alt='Coverage Status' /></a>

Please report any bugs or feature requests to C<bug-date-lectionary at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Date-Lectionary>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Date::Lectionary


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Date-Lectionary>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Date-Lectionary>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Date-Lectionary>

=item * Search CPAN

L<http://search.cpan.org/dist/Date-Lectionary/>

=back


=head1 ACKNOWLEDGEMENTS

Many thanks to my beautiful wife, Jennifer, my amazing daughter, Rosemary, and my sweet son, Oliver.  But, above all, SOLI DEO GLORIA!

=head1 LICENSE

Copyright 2016-2018 MICHAEL WAYNE ARNOLD

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

__PACKAGE__->meta->make_immutable;

1;    # End of Date::Lectionary
